### PR TITLE
feat(index): add support for `{RegExp}` directives (`options.directives`)

### PR DESCRIPTION
**Pull request by okyantoro**
Example:
The custom directive ```{name: /\?(php|=).*/, start: '<', end: '>'}``` will match any directive with ```<?php ?>``` or ```<?= ?>``` pattern.

**Note:**
I thought that this commit will solve https://github.com/posthtml/posthtml-parser/issues/26#issuecomment-358201603 if we use regular expression directive started with ```<%``` but after I checked the library, I found that any tag started with ```%``` for example ```<%=name%>``` will be treated as a tag by the library, so this tag will never touch ```onprocessinginstruction``` event. So this is an exception.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -29,11 +29,20 @@ function postHTMLParser(html, options) {
...
             var directiveText = directive.start + data + directive.end;
+            var isDirective = false;
 
-            if (name.toLowerCase() === directive.name) {
+            tagName = name.toLowerCase();
+            if ((directive.name instanceof RegExp) && directive.name.test(tagName)) {
+                isDirective = true;
+            } else if (tagName === directive.name) {
+                isDirective = true;
+            }
```
Comment (GitScrum) [inline index.js]: one condition is better:
```js
if ((directive.name instanceof RegExp) && directive.name.test(tagName)) {
    isDirective = true;
} 

if (tagName === directive.name) {
    isDirective = true;
}
```

Comment (okyantoro) [reply index.js]: fixed

File: index.js [hunk 2]
```diff
@@ -29,11 +29,22 @@ function postHTMLParser(html, options) {
         var directives = options.directives || defaultDirectives;
         var last = bufArray.last();
 
+        var tagName;
         for (var i = 0; i < directives.length; i++) {
             var directive = directives[i];
             var directiveText = directive.start + data + directive.end;
+            var isDirective = false;
```
Comment (michael-ciniawsky) [inline index.js]:
```js
function isDirective (directive, tag) {
  if (directive.name instanceof RegExp) {
    const regex = new RegExp(directive.name, 'i')

    return regex.test(tag)
  }

  if (tag !== directive.name) {
    return false
  }
  
  return true
}

File: test/test.js [hunk 3]
```diff
@@ -131,6 +131,15 @@ describe('PostHTML-Parser test', function() {
         expect(parser('<?php echo "Hello word"; ?>', customDirectives)).to.eql(['<?php echo "Hello word"; ?>']);
     });
 
+    it('should be parse regular expression directive', function() {
+        var customDirectives = {directives: [
```
Comment (michael-ciniawsky) [inline test/test.js]: `customDirectives` => `options` && `\n` after `{`

File: test/test.js [hunk 4]
```diff
@@ -131,6 +131,15 @@ describe('PostHTML-Parser test', function() {
...
+            {name: /\?(php|=).*/, start: '<', end: '>'}
```
Comment (michael-ciniawsky) [inline test/test.js]: `\s`  after `{` && `}` please

File: index.js (hunk 2 above)
Comment (michael-ciniawsky) [reply index.js]: Untested but please a make this a private helper function instead of `n` variables and `{Boolean}` switches/assignments

Comment (okyantoro) [reply index.js]: this line ```const regex = RegExp(directive.name, 'i')```
```const``` is not supported by es5 and ```RegExp(directive.name, 'i')``` only works in ES6.
Related to [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp)
> Starting with ECMAScript 6, new RegExp(/ab+c/, 'i') no longer throws a TypeError ("can't supply flags when constructing one RegExp from another") when the first argument is a RegExp and the second flags argument is present. A new RegExp from the arguments is created instead.

So we change the ```const``` to ```var``` and ```RegExp(directive.name, 'i')``` to ```RegExp(directive.name.source, 'i')```
Thank you

File: index.js [hunk 5]
```diff
@@ -33,18 +47,9 @@ function postHTMLParser(html, options) {
         for (var i = 0; i < directives.length; i++) {
             var directive = directives[i];
             var directiveText = directive.start + data + directive.end;
-            var isDirective = false;
 
             tagName = name.toLowerCase();
```
Comment (michael-ciniawsky) [inline index.js]: Just reassign `name` here

File: index.js (hunk 2 above)
Comment (michael-ciniawsky) [reply index.js]:
```
var pattern = /a/

var regex = RegExp(pattern, 'i')

console.log(regex.test('Abc'))
```

⚠️ Note that it's `RegExp(pattern, flags)` **not** `new RegExp(string, flags)` (No `new`)

File: index.js [hunk 6]
```diff
@@ -25,6 +25,20 @@ function postHTMLParser(html, options) {
         return this[this.length - 1];
     };
 
+    function isDirective(directive, tag) {
+        if (directive.name instanceof RegExp) {
+            var regex = RegExp(directive.name.source, 'i');
```
Comment (michael-ciniawsky) [inline index.js]: Final nitpick `directive.name.source` => `directive.name` (But both work)

**Comment by michael-ciniawsky** [comment]
@GitScrum Is this really a semver major ? :)

File: index.js (hunk 6 above)
Comment (okyantoro) [reply index.js]: travis gives this error if we give regex parameter for regex and supply flag.
```
1) PostHTML-Parser test should be parse regular expression directive:
     TypeError: Cannot supply flags when constructing one RegExp from another
      at new RegExp (native)
      at RegExp (native)
      at isDirective (index.js:9:991)
      at Object.parserDirective [as onprocessinginstruction] (index.js:9:2022)
      at Parser.onprocessinginstruction (node_modules/htmlparser2/lib/Parser.js:275:13)
      at Tokenizer._stateInProcessingInstruction (node_modules/htmlparser2/lib/Tokenizer.js:345:13)
      at Tokenizer._parse (node_modules/htmlparser2/lib/Tokenizer.js:686:9)
      at Tokenizer.write (node_modules/htmlparser2/lib/Tokenizer.js:632:7)
      at Parser.write (node_modules/htmlparser2/lib/Parser.js:334:18)
      at postHTMLParser (index.js:9:5732)
      at parser (index.js:9:6155)
      at parserWrapper (index.js:9:6665)
      at Context.<anonymous> (test/test.js:143:16)
```
I have tried in https://github.com/posthtml/posthtml-parser/commit/f5249b2b9b6f002ba8c68bd600867e5b73b92a7f and the test result https://travis-ci.org/posthtml/posthtml-parser/builds/330561437 is failed.

Comment (michael-ciniawsky) [reply index.js]: 🙇 I misread the MDN snippet which clearly states **starting** from ECMAScript 2015 (6) it no long throws. You're right for the current targets this needs definitely to be handled via `pattern.source`

Comment (okyantoro) [reply index.js]: Is there any plan to move to ES6?

Comment (GitScrum) [reply index.js]: we support the node> = 0.10

**Comment by GitScrum** [comment]
@okyantoro thanks for your contribution, we really appreciate this and are looking forward to a new pool of requisites 👍

**Comment by okyantoro** [comment]
Thank you @GitScrum

**Pull request merged**